### PR TITLE
[Refactor] Refactor the code of vec stream load

### PR DESCRIPTION
**Pull request by HappenLee**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by zbtzbtzbt — COMMENTED** [review]
review part

File: be/src/olap/memtable.cpp [hunk 1]
```diff
@@ -145,38 +126,33 @@ void MemTable::insert(const vectorized::Block* block, size_t row_pos, size_t num
...
 
-void MemTable::insert_one_row_from_block(RowInBlock* row_in_block_ptr)
-{
+void MemTable::_insert_one_row_from_block(RowInBlock* row_in_block) {
     _rows++;
     bool overwritten = false;
-    if (_keys_type == KeysType::DUP_KEYS)
-    {
-        _vec_skip_list->Insert(row_in_block_ptr, &overwritten);
+    if (_keys_type == KeysType::DUP_KEYS) {
```
Comment (zbtzbtzbt) [inline be/src/olap/memtable.cpp:137]: review partial code

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 2]
```diff
@@ -156,6 +157,14 @@ Status SegmentWriter::append_block(const vectorized::Block* block, size_t row_po
     return Status::OK();
 }
 
+int64_t SegmentWriter::max_row_to_add(size_t row_avg_size_in_bytes) {
+    int64_t size_rows = ((int64_t)MAX_SEGMENT_SIZE - (int64_t)estimate_segment_size()) / row_avg_size_in_bytes;
+    int64_t count_rows = (int64_t)_max_row_per_segment - _row_count ;
```
Comment (BiteTheDDDDt) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:162]:
```suggestion
    int64_t count_rows = (int64_t)_max_row_per_segment - _row_count;
```

File: be/src/runtime/thread_mem_tracker_mgr.h [hunk 3]
```diff
@@ -103,7 +103,7 @@ class ThreadMemTrackerMgr {
     void update_tracker_id(int64_t tracker_id);
 
     void add_tracker(const std::shared_ptr<MemTracker>& mem_tracker) {
-        DCHECK(_mem_trackers.find(mem_tracker->id()) == _mem_trackers.end());
+        // DCHECK(_mem_trackers.find(mem_tracker->id()) == _mem_trackers.end());
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/runtime/thread_mem_tracker_mgr.h:106]: this code of xinyi refactor of memtracker. which cause DCHECK failed, xinyi fix it in master.

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 4]
```diff
@@ -110,61 +106,27 @@ OLAPStatus BetaRowsetWriter::add_block(const vectorized::Block* block) {
...
-                return OLAP_ERR_WRITER_DATA_WRITE_ERROR;
-            }
+    do {
+        auto max_row_add = _segment_writer->max_row_to_add(row_avg_size_in_bytes);
+        if (UNLIKELY(max_row_add < 1)) {
+            // no space for another signle row, need flush now
+            RETURN_NOT_OK(_flush_segment_writer(&_segment_writer));
+            RETURN_NOT_OK(_create_segment_writer(&_segment_writer));
+            DCHECK(_segment_writer->max_row_to_add(row_avg_size_in_bytes) > 0);
+            continue;
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/olap/rowset/beta_rowset_writer.cpp:117]: yes

File: be/src/olap/memtable.h [hunk 5]
```diff
@@ -96,25 +95,19 @@ class MemTable {
...
             }
         }
     };
+
     class RowInBlockComparator {
     public:
-        RowInBlockComparator(const Schema* schema):_schema(schema){};
+        RowInBlockComparator(const Schema* schema) : _schema(schema) {};
         //call set_block before operator().
         //在第一次insert block时创建的 _input_mutable_block, 所以无法在Comparator的构造函数中获得pblock
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/olap/memtable.h:109]: ok

File: be/src/olap/memtable.cpp [hunk 6]
```diff
@@ -145,38 +126,33 @@ void MemTable::insert(const vectorized::Block* block, size_t row_pos, size_t num
     _mem_tracker->consume(newsize - oldsize);
 
     for(int i = 0; i < num_rows; i++){       
-        RowInBlock* row_in_block_ptr = new RowInBlock(cursor_in_mutableblock + i);
-        rowInBlocks.push_back(row_in_block_ptr);
-        insert_one_row_from_block(row_in_block_ptr);
+        _row_in_blocks.emplace_back(new RowInBlock{cursor_in_mutableblock + i});
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/olap/memtable.cpp:129]: all optimization of stream vec load will do after refactor

File: be/src/olap/memtable.cpp [hunk 7]
```diff
@@ -239,51 +215,40 @@ void MemTable::_aggregate_two_row(const ContiguousRow& src_row, TableKey row_in_
...
         auto place = row_in_skiplist->_agg_places[cid];
-
         auto col_ptr = _input_mutable_block.mutable_columns()[cid].get();
-
-        _agg_functions[cid]->add(place, 
-                const_cast<const doris::vectorized::IColumn**>( &col_ptr),
-                new_row->_row_pos,
-                nullptr
-                );
+        _agg_functions[cid]->add(place, const_cast<const doris::vectorized::IColumn**>( &col_ptr),
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/olap/memtable.cpp:231]: ok

File: be/src/olap/memtable.cpp (hunk 6 above)
Comment (wangbo) [inline be/src/olap/memtable.cpp:129]: Why not allocate RowInBlock in stack here?

File: be/src/olap/memtable.cpp [hunk 8]
```diff
@@ -145,38 +126,33 @@ void MemTable::insert(const vectorized::Block* block, size_t row_pos, size_t num
...
-void MemTable::insert_one_row_from_block(RowInBlock* row_in_block_ptr)
-{
+void MemTable::_insert_one_row_from_block(RowInBlock* row_in_block) {
     _rows++;
     bool overwritten = false;
-    if (_keys_type == KeysType::DUP_KEYS)
-    {
-        _vec_skip_list->Insert(row_in_block_ptr, &overwritten);
+    if (_keys_type == KeysType::DUP_KEYS) {
+        _vec_skip_list->Insert(row_in_block, &overwritten);
```
Comment (wangbo) [inline be/src/olap/memtable.cpp:138]: How about add a rethink here?
Dup table just need to sort? Is skip list the best way to sort?

File: be/src/olap/memtable.cpp [hunk 9]
```diff
@@ -145,38 +126,33 @@ void MemTable::insert(const vectorized::Block* block, size_t row_pos, size_t num
...
+        for (auto cid = _schema->num_key_columns(); cid < _schema->num_columns(); cid++){
             auto col_ptr = _input_mutable_block.mutable_columns()[cid].get();
-            auto place = row_in_block_ptr->_agg_places[cid];
-            _agg_functions[cid]->add(place, 
-                    const_cast<const doris::vectorized::IColumn**>( &col_ptr),
-                    row_in_block_ptr->_row_pos,
-                    nullptr
-                    );
+            auto place = row_in_block->_agg_places[cid];
+            _agg_functions[cid]->add(place, const_cast<const doris::vectorized::IColumn**>( &col_ptr),
```
Comment (wangbo) [inline be/src/olap/memtable.cpp:151]:
```suggestion
            _agg_functions[cid]->add(place, const_cast<const doris::vectorized::IColumn**>(&col_ptr),
```

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**